### ISSUE TITLE
Padding arrays with NaNs

### ISSUE THREAD

**Issue by TomNicholas** [zarr-developers/VirtualiZarr#22]
We want to be able to pad arrays with `NaN` values (/ zarr array `fill_value`) because some kerchunk use cases involve padding data (e.g. satellite swathes that produce slightly differently shaped data each day such as NASA SWOT).

**Comment by TomNicholas** [comment]
We should probably be able to make the xarray `.pad` method work, at least with very specific arguments, e.g. `ds.pad(x=1, mode='constant')`. We would then intercept the arguments and re-interpret the padding operation for however the underlying manifest actually works.

**Comment by dcherian** [comment]
> satellite swathes that produce slightly differently shaped data

How will this work, since each chunk will require a different padding (presumably)? Are you going to add a new codec?

For example consider three arrays with shapes (1,1), (1,2), (1,3), stacked along axis=0.

**Comment by ayushnag** [comment]
I did use kerchunk for SWOT but I wasn't able to pad them so I only used the granules with the most common swath length. This was for a demo though and ideally there is some way to pad with nan's.

**Comment by TomNicholas** [comment]
> For example consider three arrays with shapes `(1,1), (1,2), (1,3)`, stacked along axis=0.

To be concrete, if the chunksize for all these three arrays is `(1,1)`, then empty(/nonexistent) chunks can be padded to each array in turn no problem such that the shapes match and the stacking becomes possible. I agree that if the chunksizes are actually `(1,1), (1,2), (1,3)` then it's a lot harder.

**Comment by dcherian** [comment]
> I agree that if the chunksizes are actually (1,1), (1,2), (1,3) then it's a lot harder.

IIRC this is a good model for the swath problem. I think the (only?) solution is a new codec since we need to transform the chunk itself.  Something like:
```python
chunk: zarr.Array
out = np.full_like(new_shape, fill_value=chunk.fillvalue)
slicer = tuple(slice(size) for size in chunks.shape)
out[slicer] = np.array(chunk)
```

**Comment by TomNicholas** [comment]
> think the (only?) solution

I guess in this case even variable-length chunks wouldn't save us because this would actually require "ragged" chunks (i.e. no longer a rectilinear array of chunks)...

> IIRC this is a good model for the swath problem. I think the (only?) solution is a new codec since we need to transform the chunk itself. Something like:

I don't really understand what that code snippet is meant to do but my understanding of your suggestion is to add a codec which records the new padded shape of the chunk, then so when the chunk is read if you're indexing into the unpadded part it indexes into the normal zarr array, but if you're indexing into the padded part it just returns the fillvalue. But isn't the problem with that idea that you would need only certain chunks to have that codec, but codecs are set on a per-array basis, not a per-chunk basis?

**Comment by dcherian** [comment]
> I guess in this case even variable-length chunks wouldn't save us because this would actually require "ragged" chunks

Yes exactly.

> my understanding of your suggestion is to add a codec which records the new padded shape of the chunk

The "new padded shape" is just the regularized chunk shape, it should be inferred from the array metadata.

> when the chunk is read if you're indexing into the unpadded part it indexes into the normal zarr array, but if you're indexing into the padded part it just returns the fillvalue. 

At read time, all chunks are processed to a regular chunk shape so the rest of Zarr does not care about what it is actually indexing.

**Comment by mliukis** [comment]
@TomNicholas Do you happen to have an update or a timeline for when this issue will be addressed in VirtualiZarr?

**Comment by TomNicholas** [comment]
@mliukis I have not thought at all about this issue, nor am I planning to work on it in the short term. But you or others are welcome to!

> The "new padded shape" is just the regularized chunk shape, it should be inferred from the array metadata.

@dcherian re-reading this, I'm still confused as to how the codec would know what chunk shape to use at read-time. IIUC the codec doesn't have access to that information. Are you suggesting that the chunk shape be serialized into the codec parameters at write time instead?

**Comment by betolink** [comment]
> chunk shape be serialized into the codec parameters at write time instead?

That would be neat but I wonder if this itself can be reliable when there is so much inconsistent data across even the same collection. I wonder if it can be up to the users (data producers) e.g. I know this should be the chunk size, shape... therefore the padding should be X,Y,Z. 

We (ITS_LIVE) can take a initial stab at this, the idea would be to implement it in virtualizarr, in a way that will be transparent to Xarray. cc @mliukis @alex-s-gardner

**Comment by dcherian** [comment]
>  Are you suggesting that the chunk shape be serialized into the codec parameters at write time instead?

Yes, it's a parameter recorded in the Codec definition. So for `(1,1); (1,2); (1,3)` you store `final_shape: tuple[int, ...] = (1,3)` (a choice --- any number ≥ 3 works). What seems harder to me is knowing that particular chunk is `(1,2)` while a different one is `(1,3)`. Do you have access to netCDF info at run time?
